File: Graph/acmicpc.net/16929/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

class Main {
    static final String inputFilePath = "./Graph/acmicpc.net/16929/sample.txt";
    static int[][] dist;
    static char[][] a;
    static boolean[][] check;
    static int[] dx = { -1, 0, 0, 1 };
    static int[] dy = { 0, -1, 1, 0 };
    static int M, N;

    public static void main(String[] args) throws Exception {
        System.setIn(new FileInputStream(inputFilePath));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        dist = new int[N][M];
        a = new char[N][M];
        check = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            String line = br.readLine();
            for (int j = 0; j < M; j++) {
                a[i][j] = line.charAt(j);
            }
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (check[i][j])
                    continue;
                boolean ok = dfs(i, j, -1, -1, a[i][j]);
                if (ok) {
                    System.out.println("Yes");
                    System.exit(0);
                }
            }
        }
        System.out.println("No");
    }

    public static boolean dfs(int x, int y, int px, int py, char color) {
        if (check[x][y]) {
            return true;
        }
        check[x][y] = true;
        for (int i = 0; i < 4; i++) {
            int nx = dx[i] + x;
            int ny = dy[i] + y;
            if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                if (!(nx == px && ny == py) && a[nx][ny] == color) {
                    if (dfs(nx, ny, x, y, color)) {
                        return true;
                    }
                }
            }
        }

        return false;
    }
}

File: Graph/acmicpc.net/16929/Main.js
const inputFilePath = "/sample.txt";
let input = require("fs").readFileSync(__dirname + inputFilePath).toString().trim().split("\n");
// let input = require("fs").readFileSync('/dev/stdin').toString().trim().split("\n");
let lineIdx = 0;
const [DX, DY] = [[0, -1, 1, 0], [-1, 0, 0, 1]];
const [N, M] = input[lineIdx++].split(" ").map(Number);
const list = Array.from(Array(N), () => []);
const visited = Array.from(Array(N), () => Array(M).fill(0));
for (let i = 0; i < N; i++) {
  list[i] = input[lineIdx++].split("");
}
for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    if (!visited[i][j]) dfs(i, j, 1);
  }
}
console.log("No");

function dfs(x, y, depth) {
  if (!visited[x][y]) visited[x][y] = depth;
  for (let i = 0; i < 4; i++) {
    const [nx, ny] = [x + DX[i], y + DY[i]];
    if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
    if (list[x][y] === list[nx][ny]) {
      if (visited[nx][ny] === 0) {
        dfs(nx, ny, depth + 1);
      } else {
        if (visited[x][y] - visited[nx][ny] >= 3) {
          console.log("Yes");
          process.exit();
        }
      }
    }
  }
}